.github/workflows/ci-lint.yml

```diff
@@ -0,0 +1,33 @@
+name: pre-commit
+
+on:
+  pull_request:
+  push:
+     branches: [master]
+
+jobs:
+  pre-commit:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v4.1.7
+      with:
+        # Ensure the full history is fetched
+        # This is required to run pre-commit on a specific set of commits
+        # TODO: Remove this when all the pre-commit issues are fixed
+        fetch-depth: 0
+    - uses: actions/setup-python@v5.1.1
+      with:
+        python-version: 3.9
+    - name: Determine commit range
+      id: commit_range
+      run: |
+       echo "TO_REF=${{ github.sha }}" >> $GITHUB_ENV
+        if [ "${{ github.event_name }}" == "pull_request" ]; then
+          echo "FROM_REF=${{ github.event.pull_request.base.sha }}" >> $GITHUB_ENV
+        else
+          echo "FROM_REF=${{ github.event.before }}" >> $GITHUB_ENV
+        fi
+    - uses: pre-commit/action@v3.0.1
+      with:
+        # TODO: Remove this when all the pre-commit issues are fixed
+        extra_args: --from-ref ${{ env.FROM_REF }} --to-ref ${{ env.TO_REF }}
```

Comment:
Do we need this? Seems like all pre-commit issues will be fixed in this PR.

Might also be good to add ruff rules, maybe as a follow up? They will likely catch a LOT of stuff since they haven't been implemented here before...

Reply:
No I don't think so. Thanks for catching that!

I think this should include the default rules for ruff. Are there any we want to change or add?

Reply:
Like I said though these generate a lot of changes, so may not all be appropriate for a single PR.

I'd also suggest the `ruff format` pre-commit job as well; there are many tensorflow projects with mixed tabs and spaces or 2-space indents, and it would be good to get these all into shape.

Reply:
Sounds good. I added a ruff.toml and most of the included rules. There were a few changes but I think its still reasonable for this PR. I did add some extra ignores that I think should be ignored in the future.